### ISSUE TITLE
cquery returns rule_input targets which are missing from deps list

### ISSUE THREAD

**Issue by illicitonion** [bazelbuild/bazel#14617]
### Description of the problem / feature request:

Chasing the chain of `rule_input`s from running `cquery` on a `java_test` target, I see the following chain:

```
//java/example:ExampleTest
->@bazel_tools//tools/cpp:toolchain
->@local_config_cc//:toolchain
->@local_config_cc//:cc-compiler-armeabi-v7a
->@bazel_tools//tools/objc:host_xcodes
```

In particular, here's where I see `@bazel_tools//tools/objc:host_xcodes`:

```
% bazel cquery '@local_config_cc//:cc-compiler-armeabi-v7a' --output=jsonproto 2>/dev/null | jq '.results[].target.rule.ruleInput'
[
  "@bazel_tools//tools/build_defs/cc/whitelists/parse_headers_and_layering_check:disabling_parse_headers_and_layering_check_allowed",
  "@bazel_tools//tools/build_defs/cc/whitelists/starlark_hdrs_check:loose_header_check_allowed_in_toolchain",
  "@bazel_tools//tools/cpp:interface_library_builder",
  "@bazel_tools//tools/cpp:link_dynamic_library",
  "@bazel_tools//tools/objc:host_xcodes",
  "@local_config_cc//:empty",
  "@local_config_cc//:osx_tools_armeabi-v7a",
  "@local_config_cc//:stub_armeabi-v7a"
]
[
  "@bazel_tools//tools/build_defs/cc/whitelists/parse_headers_and_layering_check:disabling_parse_headers_and_layering_check_allowed",
  "@bazel_tools//tools/build_defs/cc/whitelists/starlark_hdrs_check:loose_header_check_allowed_in_toolchain",
  "@bazel_tools//tools/cpp:interface_library_builder",
  "@bazel_tools//tools/cpp:link_dynamic_library",
  "@bazel_tools//tools/objc:host_xcodes",
  "@local_config_cc//:empty",
  "@local_config_cc//:osx_tools_armeabi-v7a",
  "@local_config_cc//:stub_armeabi-v7a"
]
[
  "@bazel_tools//tools/build_defs/cc/whitelists/parse_headers_and_layering_check:disabling_parse_headers_and_layering_check_allowed",
  "@bazel_tools//tools/build_defs/cc/whitelists/starlark_hdrs_check:loose_header_check_allowed_in_toolchain",
  "@bazel_tools//tools/cpp:interface_library_builder",
  "@bazel_tools//tools/cpp:link_dynamic_library",
  "@bazel_tools//tools/objc:host_xcodes",
  "@local_config_cc//:empty",
  "@local_config_cc//:osx_tools_armeabi-v7a",
  "@local_config_cc//:stub_armeabi-v7a"
]
```

And this target seems to exist:

```
% bazel cquery '@bazel_tools//tools/objc:host_xcodes' 2>/dev/null                                                           
@bazel_tools//tools/objc:host_xcodes (51eca89)
```

But it's not returned in `deps(//java/example:ExampleTest)` or `somepath(//java/example:ExampleTest, @bazel_tools//tools/objc:host_xcodes)`:

```
% bazel cquery 'somepath(//java/example:ExampleTest, @bazel_tools//tools/objc:host_xcodes)' 2>/dev/null

% bazel cquery 'deps(//java/example:ExampleTest)' 2>/dev/null | grep -c '@bazel_tools//tools/objc:host_xcodes'
0
```

The target _does_ show up using `query` not `cquery`, so maybe this is a duplicate of https://github.com/bazelbuild/bazel/issues/14610 where the root issue is that `rule_input`s don't get configuration filtered, but target dep edges do?

Either way, it doesn't seem like there's a way to trace "Here are the files which impact this target in the current configuration", except possibly by performing a per-target `deps(target)` query - trying to batch this into a single `deps(...)` query seems to take away the ability to filter by configuration?

### What operating system are you running Bazel on?

macOS

### What's the output of `bazel info release`?

release 4.2.1

/cc @gregestren

**Comment by aranguyen** [comment]
this is the duplicate of [1406](https://github.com/bazelbuild/bazel/issues/14610). See comment https://github.com/bazelbuild/bazel/issues/14610#issuecomment-1025523391. On the top of my head, I do not think it's possible to achieve what you want in a single cquery because cquery runs over the result of the analysis phase instead of the loading phase. The work around doing per-target deps(target) query makes sense to me. @illicitonion is this a blocker for you?

**Comment by illicitonion** [comment]
> I do not think it's possible to achieve what you want in a single cquery because cquery runs over the result of the analysis phase instead of the loading phase.

This is where I'm a bit confused - https://github.com/bazelbuild/bazel/issues/14610#issuecomment-1025523391 suggests that the problem here is that we're using Target messages from the loading phase, but as cquery uses the results of analysis, I'd expect the Target messages to be refined with the information we gained by doing further analysis. It feels like come analysis time we must _have_ (or at least be able to construct) a filtered down Target message, but we happen to be re-using the one from the loading phase? So the real request here is to refine the returned Target using the information we gained by doing further analysis. Does that make sense as a problem statement?

> The work around doing per-target deps(target) query you mentioned makes sense to me.

Yeah, it's a functional workaround, but as we're doing whole-repo queries, that necessitates doing many thousands of individual queries, and the overhead of a single bazel invocation to do a query on a pre-warmed graph is hundreds of milliseconds, so doing per-target queries would add hours to the query pipeline, which is definitely not a usable solution.

> @illicitonion is this a blocker for you?

Yeah, this is a pretty hard blocker - right now we're both over-estimating which targets we need to run (by following edges which don't have relevant configurations), and ignoring errors (because if things are missing from the query, we have to assume they were `select`ed out, rather than that something went wrong in the graph structure or our parsing logic).

It would be great to discuss whether there are important reasons for us to be returning the loading-phase `Target`, or whether we could refine that down to a post-configuration `Target` with more precise information. I'm happy to dig into the code and try to put together a fix, with a little mentoring...

Thanks!
